Tweaks for map generator performance.

RoosterDragon opened the pull request:
The map generator is already performant, but we can remove a few rough edges from infrastructure it interacts with.

- Add TypeDictionary constructor to clone from an existing copy. The map editor infrastructure relies heavily on cloning ActorReferences and the underlying dictionaries. A dedicated clone method avoids reflection overhead for CreateTypeContainer on a fresh instance.
- Use ThrowIndexOutOfRangeException helper. This allows the Index method to be inlined.
- Reuse lists in FloodFill. This avoids allocating and resizing new lists on each loop when we can reuse the existing capacity.
- Manage CellEntryChanged at the layer rather than individual previews. We only need to attach/detach from the delegate once at the layer, rather than many times as previews are added and removed.

RoosterDragon committed:
Tweaks for map generator performance.

The map generator is already performant, but we can remove a few rough edges from infrastructure it interacts with.

- Add TypeDictionary constructor to clone from an existing copy. The map editor infrastructure relies heavily on cloning ActorReferences and the underlying dictionaries. A dedicated clone method avoids reflection overhead for CreateTypeContainer on a fresh instance.
- Use ThrowIndexOutOfRangeException helper. This allows the Index method to be inlined.
- Reuse lists in FloodFill. This avoids allocating and resizing new lists on each loop when we can reuse the existing capacity.
- Manage CellEntryChanged at the layer rather than individual previews. We only need to attach/detach from the delegate once at the layer, rather than many times as previews are added and removed.

pchote commented on OpenRA.Game/Primitives/TypeDictionary.cs:
```diff
@@ -20,7 +21,17 @@ public class TypeDictionary : IEnumerable<object>
+
+		public TypeDictionary(TypeDictionary cloneFrom)
+		{
+			data = new Dictionary<Type, ITypeContainer>(cloneFrom.data.Select(kvp => KeyValuePair.Create(kvp.Key, kvp.Value.Clone())));
```
.ToDictionary()?

PunkPun replied:
```suggestion
			data = cloneFrom.data.ToDictionary(kvp => kvp.Key, kvp => kvp.Value.Clone());
```

PunkPun commented:
@ashleynewson ping

ashleynewson commented:
Did you already do any benchmarks? If not, I could try running it through the fuzzer. :)

ashleynewson commented on OpenRA.Game/Map/ActorReference.cs:
```diff
@@ -54,13 +54,8 @@ public ActorReference(string type, Dictionary<string, MiniYaml> inits)
-				return dict;
-			});
+			var initsClone = new TypeDictionary(inits);
+			initDict = Exts.Lazy(() => initsClone);
```
If we're using `Lazy`, would it make more sense to do the `new TypeDictionary(inits)` within the `Lazy` lambda?

PunkPun replied:
The original data could have changed? So no?

pchote replied:
The original code lazily read the original data too, so inside the lambda preserves behaviour (whether or not that makes sense is a different question...)

RoosterDragon replied:
It's definitely a fair spot that this changes behaviour compared to before.

The laziness in ActorReference is intended for the main constructor where we don't want to call `LoadInit` until some other global state (🤮) is in a good place.

It's definitely not intended to support a scenario where you send a dictionary but expect to mutate it before you evaluate the ActorReference - I think that'd be pretty cursed. I've reviewed the callers and thankfully they're all just passing throw-away dicts so there's no observable impact in behaviour to make it eager instead and remove the curse here.

I've pushed an additional commit which fixes the other ctor to avoid this pitfall as well.

pchote replied:
> The laziness in ActorReference is intended for the main constructor where we don't want to call `LoadInit` until some other global state (🤮) is in a good place.

I have prototyped a fix for this (https://github.com/pchote/OpenRA/commit/db0df1a4f95930c73233dc011e8e076ad5be9c2b) which I plan to turn into a PR after a couple more prereqs are merged.

ashleynewson commented on OpenRA.Mods.Common/MapGenerator/MatrixUtils.cs:
```diff
@@ -345,11 +345,12 @@ public static void FloodFill<P>(
 			Func<int2, P, P?> filler,
 			ImmutableArray<int2> spread) where P : struct
 		{
+			var current = new List<(int2 XY, P Prop)>(spread.Length);
```
`spread.Length` probably isn't a good initial capacity. That's only going to be 4 or 8. Typically, it will be allocated in multiples of that. somewhat proportional to the length of `seeds`.

In the absence of profiling data, I'd probably just use the non-length constructor. If we did want to use a length hint, I suspect that `seeds.Length` would make more sense. Though, it could also be the case that having both `next` and `current` hold an initial capacity of `seeds.Length * spread.Length` makes sense.

Note that with each iteration, the number of `next`/`current` points may either grow or shrink. It depends on how the flood fill is being used.

ashleynewson replied:
From a quick benchmark, it doesn't look like the initial capacity has much influence on performance.

RoosterDragon replied:
I'll leave it blank for now.

ashleynewson commented on OpenRA.Mods.Common/MapGenerator/MatrixUtils.cs:
```diff
@@ -345,11 +345,12 @@ public static void FloodFill<P>(
 			{
-				var current = next;
-				next = [];
+				(next, current) = (current, next);
```
There is also `CellLayerUtils.FloodFill`, which can also be updated.

ashleynewson commented:
I did some benchmarks with and without patch using the fuzzer. Note that this only measures the map generator itself, and not any of the associated map editor logic such as EditorActorLayer. I think this would still be largely representative of lobby map generation though!

Bash command:
```
time ./utility.sh ra --fuzz-map-generator --generator=experimental --tilesets=TEMPERAT,DESERT,SNOW --sizes=64x64,96x96,128x128 --choices=Seed="$(seq -s, 100)"
```

Without patch:
```
real	6m16.715s
user	6m35.350s
sys	0m8.308s
```

With patch:
```
real	5m18.507s
user	5m33.992s
sys	0m6.321s
```

So, a decent 15% reduction to the inherent map generation time on my system. :+1:

RoosterDragon committed:
ActorReference takes MiniYaml as input instead of a Dict.

As the InitDict is lazy, this removes the possibility for a mutable collection to be used but the values not to be consumed until later, which would be confusing. As all existing callers actually create a throw-away dictionary from MiniYaml, this is an easy switch.

PunkPun reviewed (approved):
Ok

PunkPun commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/a1b1653047cb19b0d0904f02ceb9e864f5661ec6...dad4a1fb08183df67e8831a654355112ac232b12
